plutus-scb/src/Cardano/Protocol/Socket/Client.hs

```diff
@@ -0,0 +1,203 @@
+              _ <- runNodeClientEffects connection clientPrj $
+                resumeLocalState point
+              return requestNext
+        , ChainSync.recvMsgIntersectNotFound = error "Not supported."
+        }
+
+      requestNext :: ChainSync.ClientStIdle Block Tip IO ()
+      requestNext =
+        ChainSync.SendMsgRequestNext
+          handleNext
+          (return handleNext)
+
+      handleNext :: ChainSync.ClientStNext Block Tip IO ()
+      handleNext =
+        ChainSync.ClientStNext
+        {
+          ChainSync.recvMsgRollForward  = \block _ ->
+            ChainSync.ChainSyncClient $ do
+              _ <- runNodeClientEffects connection clientPrj
+                processBlock
+              return requestNext
+        , ChainSync.recvMsgRollBackward = error "Not supported."
+        }
+
+txSubmissionClient :: Connection
+                   -> MVar ClientStreamProjection
+                   -> TQueue Tx
+                   -> TxSubmission.LocalTxSubmissionClient Tx String IO ()
+txSubmissionClient connection clientPrj txInput =
+    TxSubmission.LocalTxSubmissionClient pushTxs
+    where
+      pushTxs :: IO (TxSubmission.LocalTxClientStIdle Tx String IO ())
+      pushTxs = do
+        header <- atomically $ readTQueue txInput
+        _ <- runNodeClientEffects connection clientPrj $
+          queueTx header
+        return $ TxSubmission.SendMsgSubmitTx
+                   header
+                   (const pushTxs) -- ignore rejects for now
+
+type NodeClientM = (ReaderT Connection (LoggingT IO))
+
+runNodeClientEffects ::
+    Connection
+ -> MVar ClientStreamProjection
+ -> Eff (NodeClientEffects NodeClientM) a
+ -> IO a
+runNodeClientEffects connection st action = do
+  prj            <- liftIO $ takeMVar st
+  (((result, ncEvents), cEvents), prj') <- liftIO
+    $ runStderrLoggingT
+    $ flip runReaderT connection
+    $ Eff.runM
+    $ Eff.runStderrLog
+    $ Eff.runState prj
+    $ handleEventLog
+    $ Eff.runWriter
+    $ Eff.runWriter
+    -- Note: The chain state in the node client *should not* be
+    --       updated directly through the state, but through
+    --       refreshing the projection to avoid cases where we directly
+    --       modify the projection in inconsistent ways.
+    --       As it is now, the algebra can cause unexpected behaviors, I think.
+    --
+    --       .. In the current code, it is not updated directly, since the
+    --       only instance  where that happens is when calling `processBlock` which is
+    --       a control call for the mocked node server (and we are interpreting for
+    --       the node client).
+    --       But we can write an algebra term that would trigger this behavior
+    --       and everything will happily compile.
+    --
+    -- Meta: Is it really a good idea to have an algebra that is *partially* interpreted
+    --       in different components? Maybe we should decide not to have that and audit
+    --       the code to find such instances.
```

Comment:
I am not sure if this is a bug, but to me, it looks like one.

Reply:
Yeah it does look like a problem. Unfortunately I haven't fixed it yet (I've been concentrating on the `TestApp` stuff). At first glance I would say that `handleChain'` should not need the `State` effect. So we probably need two different handlers for each effect - one that works with the `eventful` event log, and one that works with the state, like it's done in the emulator. What do you think?

Reply:
I don't think it needs it either. I also think that having it may be a bug since it would allow any program access to the projection, possibly modifying it in weird ways.